File: apps/web/src/game/canonical.ts
// Canonical JSON serialization and content hashing -- a TypeScript port of
// packages/contracts/src/networked_players_contracts/canonical.py, not a
// structurally-similar reimplementation: both must produce byte-identical
// canonical strings for the same value (keys sorted at every nesting level,
// no insignificant whitespace, non-ASCII left as literal UTF-8 rather than
// escaped) so a content hash computed in either language agrees. See ADR
// 0043's corrective-slice-4.6 addendum.
//
// Used client-side to verify a fetched round's `round_content_fingerprint`
// against what a frozen daily-manifest entry expects, without trusting the
// server not to have silently changed the round underneath an already-
// shared date.

/** A deterministic JSON string: keys sorted at every nesting level, no
 * insignificant whitespace. Matches Python's
 * `json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)`. */
export function canonicalJson(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(canonicalJson).join(",")}]`;
  }
  const record = value as Record<string, unknown>;
  const keys = Object.keys(record).sort();
  const parts = keys.map((key) => `${JSON.stringify(key)}:${canonicalJson(record[key])}`);
  return `{${parts.join(",")}}`;
}

async function sha256Hex(text: string): Promise<string> {
  const bytes = new TextEncoder().encode(text);
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/** A truncated sha256 hex digest of `value`'s canonical JSON form. Matches
 * `networked_players_contracts.canonical.content_hash`. Async: browsers only
 * expose SHA-256 via the (async) SubtleCrypto API. */
export async function contentHash(value: unknown, length = 16): Promise<string> {
  const digest = await sha256Hex(canonicalJson(value));
  return digest.slice(0, length);
}

/** Matches `networked_players_graph_core.connection_rounds
 * ::round_content_fingerprint` / the dependency-free mirror in
 * `networked_players_contracts.connection_rounds`. */
export async function roundContentFingerprint(round: unknown): Promise<string> {
  return `rfp-${await contentHash(round, 16)}`;
}

File: packages/contracts/src/networked_players_contracts/canonical.py
"""Canonical JSON serialization and content hashing, shared verbatim between
generation-time code (`networked_players_graph_core`) and the dependency-free
validators in this package -- not a structurally-mirrored duplicate, an
actual shared import, so the two can never silently drift on what "the same
content" means (see ADR 0043's corrective-slice-4.6 addendum).

Deterministic and insensitive to insignificant JSON formatting: key order,
whitespace, and which serializer happened to write a file to disk never
change the result -- only the actual value tree does. Both Python's
`json.dumps` here and the TypeScript port in `apps/web/src/game/canonical.ts`
must produce byte-identical canonical strings for the same value (sorted
keys, no extra whitespace, non-ASCII characters left unescaped rather than
`\\uXXXX`-encoded) so a content hash computed in either language agrees.
"""

from __future__ import annotations

import hashlib
import json
from typing import Any


def canonical_json(value: Any) -> str:
    """A deterministic JSON string: keys sorted at every nesting level, no
    insignificant whitespace, non-ASCII left as literal UTF-8 (matching
    JavaScript's `JSON.stringify` default) so this agrees byte-for-byte with
    the TypeScript port for the same value."""
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def content_hash(value: Any, *, length: int = 16) -> str:
    """A truncated sha256 hex digest of `value`'s canonical JSON form, hashed
    as UTF-8 bytes (matching `TextEncoder().encode(...)` in the TypeScript
    port)."""
    digest = hashlib.sha256(canonical_json(value).encode("utf-8")).hexdigest()
    return digest[:length]


def stable_id_digest(*parts: str, length: int = 10) -> str:
    """A truncated sha256 hex digest of pipe-joined string `parts` -- the
    primitive behind a round's own stable, content-derived id
    (`connection_rounds.py::_stable_id`). Shared here so a dependency-free
    validator can recompute a round's id from its own published semantic
    fields (endpoints + accepted answers) and prove self-consistency without
    needing the original Discogs credit graph."""
    return hashlib.sha256("|".join(parts).encode()).hexdigest()[:length]
